### ISSUE TITLE
Running gcovr in docker image 

### ISSUE THREAD

**Issue by lambtonr** [gcovr/gcovr#598]
i have gcovr coverage reporting working in local builds, however when i run gcovr to generate a coverage report during a pipeline using a docker image i different behaviors. 

1 - 0% coverage using xml reporter, when running locally it returns 97%

```
Filters for --root: (1)
- re.compile('^/builds/git/IniFile/')
Filters for --filter: (1)
- DirectoryPrefixFilter(/builds/git/IniFile/)
Filters for --exclude: (0)
Filters for --gcov-filter: (1)
- AlwaysMatchFilter()
Filters for --gcov-exclude: (0)
Filters for --exclude-directories: (0)
Scanning directory /builds/git/IniFile for gcda/gcno files...
Found 27 files (and will process 19)
Pool started with 1 threads
Processing file: /builds/git/IniFile/build/test/tests.ltrans4.ltrans.gcno
Running gcov: 'gcov /builds/git/IniFile/build/test/tests.ltrans4.ltrans.gcno --branch-counts --branch-probabilities --preserve-paths --object-directory /builds/git/IniFile/build/test' in '/tmp/tmpiqw6rtj5'
Processing file: /builds/git/IniFile/build/test/tests.ltrans6.ltrans.gcno
Running gcov: 'gcov /builds/git/IniFile/build/test/tests.ltrans6.ltrans.gcno --branch-counts --branch-probabilities --preserve-paths --object-directory /builds/git/IniFile/build/test' in '/tmp/tmpiqw6rtj5'
Processing file: /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda
Running gcov: 'gcov /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda --branch-counts --branch-probabilities --preserve-paths --object-directory /builds/IniFile/build/src/CMakeFiles/iniFile.dir' in '/tmp/tmpiqw6rtj5'
Finding source file corresponding to a gcov data file
  currdir      /builds/git/IniFile/build
  gcov_fname   /tmp/tmpiqw6rtj5/#builds#git#IniFile#src#IniFile.cpp.gcov
               ['        -', '    0', 'Source', '/builds/git/IniFile/src/IniFile.cpp\n']
  source_fname /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda
  root         /builds/git/IniFile
  fname        /builds/git/IniFile/src/IniFile.cpp
Parsing coverage data for file /builds/git/IniFile/src/IniFile.cpp

...
Gathered coveraged data for 13 files
lines: 0.0% (0 out of 679)
branches: 0.0% (0 out of 3198)
```

2 - when generating html report i get the following error

```
Traceback (most recent call last):
  File "/usr/local/bin/gcovr", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/gcovr/__main__.py", line 280, in main
    error_occurred = print_reports(covdata, options, logger)
  File "/usr/local/lib/python3.8/dist-packages/gcovr/__main__.py", line 432, in print_reports
    if generator(covdata, output.abspath, options):
  File "/usr/local/lib/python3.8/dist-packages/gcovr/writer/html.py", line 401, in print_html_report
    lines = formatter.highlighter_for_file(data['filename'])(source_file.read())
  File "/usr/local/lib/python3.8/dist-packages/gcovr/writer/html.py", line 159, in highlighter_for_file
    from jinja2 import Markup
ImportError: cannot import name 'Markup' from 'jinja2' (/usr/local/lib/python3.8/dist-packages/jinja2/__init__.py)
```

for this command :

``` sh
gcovr --html --html-details --exclude-unreachable-branches --print-summary -o coverage.html --root ../
```

gcovr is installed in docker image using [setup-cpp](https://github.com/aminya/setup-cpp) as such: ``./setup_cpp_linux --compiler $compiler --cmake true --ninja true --conan true --ccache true --cppcheck true --gcovr true --doxygen true``

**Comment by Spacetown** [comment]
In both logs there are messages like the following:
```
(WARNING) Unrecognized GCOV output for /build/devel/tracker/src/vision/imageprocessing.cpp
	  branch  1 taken -179954
```

This indicates a gcov error, see #583.

**Comment by sergiud** [comment]
How is this related to the actual problem? Besides, I use `--gcov-ignore-parse-errors` which is supposed to ignore errors not quietly quit or reduce coverage because it is ran in a Docker image.

**Comment by Spacetown** [comment]
With the option `--gcov-ignore-parse-errors` you ignore this error and continue with the next coverage file. The whole coverage file is ignored. It seems to be a problem with memory corruption and this isn't reproducible. It can differ in the docker container.
In previous versions this lines where `covered` because they are not zero.

**Comment by latk** [comment]
@sergiud You seem to have a different issue from OP, also using a different gcovr version, also using a different set from command line options. Please consider opening a separate issue so that these problems can be solved separately.

(cc @Spacetown I've reopened this issue until OP's problems have been addressed)

**Comment by sergiud** [comment]
Sure, I can open a new issue. Thanks for considering it. However, both problems seem to be related.

**Comment by latk** [comment]
@lambtonr, you have shown how you invoked gcovr for the HTML output, but you haven't shown how you produce the Cobertura/XML report. Could you provide a bit more details here? Also, which compilers are you using locally and in your container?

**Comment by lambtonr** [comment]
@latk thanks for the info regarding the html error. 

In regard to the xml output i am running the following: 
``gcovr --xml-pretty --exclude-unreachable-branches --print-summary -o coverage.xml --root ${CI_PROJECT_DIR}`` 

i have another project exhibiting the same behavior with ``--root ../``

i have updated the host compiler was 7.5 (both xml and html working)

docker image : ubuntu:latest(20.04) --GCC 11.1.0
host : ubuntu-18.04 --GCC 11.1.0

on host i get the expected behavior: 

```sh
$ gcovr --html --html-details --exclude-unreachable-branches --print-summary -o coverage.html --root ../
lines: 90.4% (686 out of 759)
branches: 42.1% (1346 out of 3198)
$ gcovr --xml-pretty --exclude-unreachable-branches --print-summary -o coverage.xml --root ../
lines: 90.4% (686 out of 759)
branches: 42.1% (1346 out of 3198)
```

**Comment by latk** [comment]
In #597 we figured out that gcovr has issues with GCC 11 since it changes the relative locations in which .gcda/.gcno files are generated. Maybe the 0% coverage is related to that problem.

**Comment by lambtonr** [comment]
@latk i have built a new docker image with my toolchain preinstalled rather than installing it as part of the pipeline via [setup-cpp](https://github.com/aminya/setup-cpp). Everything works now both under GCC-11 and clang-13. 

 GCC `gcovr --verbose --html --html-details --exclude-unreachable-branches --print-summary -o coverage.html --root ../`
```sh
Gathered coveraged data for 13 files
lines: 90.4% (686 out of 759)
branches: 42.1% (1346 out of 3198)
```
clang `gcovr --verbose --html --html-details --exclude-unreachable-branches --print-summary -o coverage.html --root ../ --gcov-executable "llvm-cov gcov"`
```sh
Gathered coveraged data for 16 files
lines: 93.4% (1020 out of 1092)
branches: 41.8% (2444 out of 5852)
```

here is the same diagnostic information from the successful gcc11 run for comparison to my original post:

```sh
Filters for --root: (1)
- re.compile('^\\/builds\\/git\\/IniFile\\/')
Filters for --filter: (1)
- DirectoryPrefixFilter(\/builds\/git\/IniFile\/)
Filters for --exclude: (0)
Filters for --gcov-filter: (1)
- AlwaysMatchFilter()
Filters for --gcov-exclude: (0)
Filters for --exclude-directories: (0)
Scanning directory ../ for gcda/gcno files...
Found 16 files (and will process 8)
Pool started with 1 threads
Processing file: /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda
Running gcov: 'gcov /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda --branch-counts --branch-probabilities --preserve-paths --object-directory /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir' in '/tmp/tmp11qzwbgz'
Finding source file corresponding to a gcov data file
  currdir      /builds/git/IniFile/build
  gcov_fname   /tmp/tmp11qzwbgz/#builds#git#IniFile#src#IniFile.cpp.gcov
               ['        -', '    0', 'Source', '/builds/git/IniFile/src/IniFile.cpp\n']
  source_fname /builds/git/IniFile/build/src/CMakeFiles/iniFile.dir/IniFile.cpp.gcda
  root         /builds/git/IniFile
  fname        /builds/git/IniFile/src/IniFile.cpp
  ...
  ```

**Comment by aminya** [comment]
FYI, setup-cpp v0.14.3 is released now, which uses gcovr 5.1
https://github.com/aminya/setup-cpp/releases/tag/v0.14.3
